File: backend/src/models/Madrasa.js
const mongoose = require('mongoose');

const MadrasaSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true,
      trim: true,
    },
    domain: {
      type: String,
      unique: true,
      trim: true,
      lowercase: true,
    },
    contactInfo: {
      phone: {
        type: String,
      },
      email: {
        type: String,
        trim: true,
        lowercase: true,
      },
    },
    enabledModules: {
      type: [String],
      default: ['students', 'teachers', 'attendance'], // Basic starter modules enabled by default
    },
    subscription: {
      plan: {
        type: String,
        required: true,
        enum: ['Basic', 'Standard', 'Premium'],
        default: 'Basic',
      },
      expiresAt: {
        type: Date,
      },
    },
  },
  {
    timestamps: true,
  }
);

MadrasaSchema.index({ domain: 1 });
MadrasaSchema.index({ 'subscription.plan': 1, 'subscription.expiresAt': 1 });

module.exports = mongoose.model('Madrasa', MadrasaSchema);

File: frontend/lib/models/madrasa.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'madrasa.freezed.dart';
part 'madrasa.g.dart';

@freezed
class Madrasa with _$Madrasa {
  const factory Madrasa({
    required String id,
    required String name,
    String? domain,
    PhoneInfo? contactInfo,
    @Default([]) List<String> enabledModules,
    @Default('Basic') String plan,
    String? subscriptionExpiresAt,
  }) = _Madrasa;

  factory Madrasa.fromJson(Map<String, dynamic> json) => _$MadrasaFromJson(json);
}

@freezed
class PhoneInfo with _$PhoneInfo {
  const factory PhoneInfo({
    String? phone,
    String? email,
  }) = _PhoneInfo;

  factory PhoneInfo.fromJson(Map<String, dynamic> json) => _$PhoneInfoFromJson(json);
}
